wagtail/admin/mail.py

```diff
@@ -85,30 +101,214 @@ def send_notification(page_revision_id, notification, excluded_user_id):
+
+    def get_template_set(self, notifying_instance, notification, **kwargs):
+        """Return a dictionary of template paths for the templates for the email subject and the text and html
+        alternatives"""
+        template_base = self.get_template_base_prefix(notifying_instance) + notification
+
+        template_subject = 'wagtailadmin/notifications/' + template_base + '_subject.txt'
+        template_text = 'wagtailadmin/notifications/' + template_base + '.txt'
+        template_html = 'wagtailadmin/notifications/' + template_base + '.html'
+
+        return {
+            'subject': template_subject,
+            'text': template_text,
+            'html': template_html,
+        }
+
+    def get_context(self, notifying_instance, notification, **kwargs):
+        return {'settings': settings}
+
+    def send_emails(self, template_set, context, recipients, **kwargs):
+
+        connection = get_connection()
+
+        with OpenedConnection(connection) as open_connection:
+
+            # Send emails
+            sent_count = 0
+            for recipient in recipients:
+                try:
+
+                    # update context with this recipient
+                    context["user"] = recipient
+
+                    # Translate text to the recipient language settings
+                    with override(recipient.wagtail_userprofile.get_preferred_language()):
+                        # Get email subject and content
+                        email_subject = render_to_string(template_set['subject'], context).strip()
+                        email_content = render_to_string(template_set['text'], context).strip()
+
+                    kwargs = {}
+                    if getattr(settings, 'WAGTAILADMIN_NOTIFICATION_USE_HTML', False):
+                        kwargs['html_message'] = render_to_string(template_set['html'], context)
+
+                    # Send email
+                    send_mail(email_subject, email_content, [recipient.email], connection=open_connection, **kwargs)
+                    sent_count += 1
+                except Exception:
+                    logger.exception(
+                        "Failed to send notification email '%s' to %s",
+                        email_subject, recipient.email
+                    )
+
+        return sent_count == len(recipients)
+
+    def __call__(self, notifying_instance, notification, **kwargs):
+        """Send emails corresponding to the notification (eg 'approved') from an instance, notifying_instance"""
+
+        if not self.can_handle(notifying_instance, notification, **kwargs):
+            return False
+
+        recipients = self.get_valid_recipients(notifying_instance, notification, **kwargs)
+
+        if not recipients:
+            return True
+
+        template_set = self.get_template_set(notifying_instance, notification, **kwargs)
+
+        context = self.get_context(notifying_instance, notification, **kwargs)
```

Comment:
Any reason not to move this into the `for recipient in recipients` loop (so that subclasses can set up whatever context they like on a per-recipient basis, rather than `context['user']` being a one-off special case)?

Reply:
The main reason I did this was simply for efficiency - most context methods involve a few queries, and don't vary too much based on user. If we decide to vary this in future, maybe we could have a base context method outside the loop, and a specific context one inside it?